Sources/ProtobufferCoding/Encoding/Encoder.swift

```diff
@@ -0,0 +1,166 @@
+        } catch let error as ProtoValidationError {
+            // Note that in this function (the one encoding values into fields, instead of encoding entire messages), our requirements to the
+            switch error {
+            case .topLevelArrayNotAllowed:
+                // We swallow all "T cannot be a top-level type" errors, since we're not encoding T into a top-level type (but rather into a field).
+                break
+            default:
+                throw error
+            }
+        }
+        let dstBufferRef = Box(ByteBuffer())
+        let encoder = _ProtobufferEncoder(codingPath: [], dstBufferRef: dstBufferRef, context: EncoderContext())
+        if getProtoCodingKind(type(of: value)) == .message {
+            encoder.context.pushSyntax(value is Proto2Codable ? .proto2 : .proto3) // no need to pop here
+        }
+        var keyedEncoder = encoder.container(keyedBy: FixedCodingKey.self)
+        try keyedEncoder.encode(value, forKey: .init(intValue: field.fieldNumber))
+        buffer.writeImmutableBuffer(dstBufferRef.value)
+    }
+}
+
+
+class EncoderContext {
+    private var syntaxStack = Stack<ProtoSyntax>()
+    private var fieldsMarkedAsRequiredOutput: Set<[Int]> = [] // int values of the coding path to the field
+    
+    /// The current syntax.
+    /// - Note: Since proto2 and proto3 messages can be contained in each other, the value of this property can change while encoding an object.
+    var syntax: ProtoSyntax { syntaxStack.peek()! }
+    
+    init(syntax: ProtoSyntax = .proto3) {
+        syntaxStack.push(syntax)
+    }
+    
+    func pushSyntax(_ syntax: ProtoSyntax) {
+        syntaxStack.push(syntax)
+    }
+    
+    func popSyntax() {
+        precondition(syntaxStack.count >= 2, "Imbalanced pop")
+        syntaxStack.pop()
+    }
+    
+    /// Marking a coding path as "required output" will indicate to the encoder that values for this field should
+    /// always be written into the proto buffer, even if the value is the field's zero default value.
+    func markAsRequiredOutput(_ codingPath: [CodingKey]) {
+        fieldsMarkedAsRequiredOutput.insert(Self.codingPathToInts(codingPath))
+    }
+    
+    func markAsRequiredOutput(_ codingPaths: [[CodingKey]]) {
+        for codingPath in codingPaths {
+            markAsRequiredOutput(codingPath)
+        }
+    }
+    
+    func unmarkAsRequiredOutput(_ codingPath: [CodingKey]) {
+        fieldsMarkedAsRequiredOutput.remove(Self.codingPathToInts(codingPath))
+    }
+    
+    func unmarkAsRequiredOutput(_ codingPaths: [[CodingKey]]) {
+        for codingPath in codingPaths {
+            unmarkAsRequiredOutput(codingPath)
+        }
+    }
+    
+    func isMarkedAsRequiredOutput(_ codingPath: [CodingKey]) -> Bool {
+        fieldsMarkedAsRequiredOutput.contains(Self.codingPathToInts(codingPath))
+    }
+    
+    private static func codingPathToInts(_ codingPath: [CodingKey]) -> [Int] {
```

Comment:
The integers in the array aren't meant to represent the field numbers, they just happen to do so. In practice we don't care about the actual contents of the array, we just need an injective mapping function, so that we can compare whether a coding path was previously registered with the context. Will still rename it, though